components/weekly_storage/weekly_event_storage_unittest.cc

```diff
@@ -0,0 +1,121 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/weekly_storage/weekly_event_storage.h"
+
+#include <memory>
+#include <utility>
+
+#include "base/test/simple_test_clock.h"
+#include "base/time/time.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/testing_pref_service.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace {
+constexpr char kPrefName[] = "brave.weekly_event_test";
+}
+
+class WeeklyEventStorageTest : public ::testing::Test {
+ public:
+  enum TestValues {
+    kNull,
+    kFoo,
+    kBar,
+    kBrave,
+    kMaxValue = kBrave,
+  };
+
+  WeeklyEventStorageTest() {
+    pref_service_.registry()->RegisterListPref(kPrefName);
+    CreateWeeklyEventStorage();
+  }
+
+  void CreateWeeklyEventStorage() {
+    clock_ = new base::SimpleTestClock;
+    clock_->SetNow(base::Time::Now());
+    state_ = std::make_unique<WeeklyEventStorage>(
+        &pref_service_, kPrefName, std::unique_ptr<base::Clock>(clock_));
+  }
+
+ protected:
+  base::SimpleTestClock* clock_ = nullptr;
```

Comment:
why this is not a `unique_ptr`?

Reply:
n/m actually it is correct. Can just mark as `// not owned.` or something

Reply:
Nevermind. Per discussion, `WeeklyEventStorage` owns the clock, and the test just holds onto an implicitly week pointer so it can manipulate time during the test. `SimpleTestClock` is thread-safe, so the code is safe, I think.

Reply:
> `SimpleTestClock` is thread-safe, so the code is safe, I think.

a lot of things in `base` is not thread-safe, including `SimpleTestClock`, but that's okay and expected, because it's very rare when we need to access something from both threads _simultaneously_. In most cases the multi-threading aspect is handled everywhere through `PostTask`s.

The neat thing about browser tests is that a test body runs in the UI (main) thread of a browser, so we can access almost anything without breaking threads restrictions.

Reply:
Makes sense. thanks for clarifying. I was unclear if the the `SearchEngineTracker` code was running on the main thread or not.

My comment about `SimpleTestClock` was based on [this code comment](https://github.com/chromium/chromium/blob/95.0.4638.54/base/test/simple_test_clock.h#L15-L17). I didn't read the implementation but it does seem to acquire a lock inside each method?